File: src/domain/services/streamService.go
package services

import (
	"Theatrum/constants"
	"Theatrum/domain/models"
	"path"
	"strings"
)

type StreamService struct {
	pathTemplateService *PathTemplateService
}

func NewStreamService(pathTemplateService *PathTemplateService) *StreamService {
	return &StreamService{
		pathTemplateService: pathTemplateService,
	}
}

func (s *StreamService) GetStreamStoragePath(stream *models.Stream, templatingVars map[string]string) (string, error) {
	// If quality is not provided, use the default quality
	if _, ok := templatingVars["quality"]; !ok {
		templatingVars["quality"] = constants.DefaultQuality
	}

	streamStorageTemplate := stream.Path

	// Only HLS-only mode uses quality subdirs.
	// DASH-enabled modes (DASH-only or dual) use a flat layout — no quality subdir appended.
	hlsOnly := stream.Distribution.HlsEnabled() && !stream.Distribution.DashEnabled()

	isRootResource := templatingVars["resource"] == constants.MasterPlaylist ||
		templatingVars["resource"] == constants.DashManifest

	if hlsOnly && !isRootResource && !strings.Contains(stream.Path, constants.PlaceholderBegin+"quality"+constants.PlaceholderEnd) {
		streamStorageTemplate += "/" + templatingVars["quality"]
	}

	// Replace the placeholders in the path
	storagePath, err := s.pathTemplateService.ReplacePlaceholders(streamStorageTemplate, templatingVars)
	if err != nil {
		return "", err
	}

	// Add the video directory to the path
	finalStoragePath := path.Join(constants.VideoDir, storagePath)

	return finalStoragePath, nil
}

File: src/main/java/com/fjourdren/theatrum/domain/service/StreamService.java
package com.fjourdren.theatrum.domain.service;

import com.fjourdren.theatrum.application.port.in.ServeStreamUseCase;
import com.fjourdren.theatrum.domain.constant.TemplateConstants;
import com.fjourdren.theatrum.domain.constant.VideoConstants;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.Stream;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/** Resolves where a stream's files live on disk. */
@Component
@RequiredArgsConstructor
public class StreamService implements ServeStreamUseCase {

    /** Resources that live at the stream root, never inside a quality subdirectory. */
    private static final Set<String> ROOT_RESOURCES = Set.of(
            VideoConstants.MASTER_PLAYLIST, VideoConstants.DASH_MANIFEST, VideoConstants.THUMBNAIL_FILE);

    private final PathTemplateService pathTemplateService;
    private final AppPaths appPaths;

    /**
     * Resolves the on-disk directory for a stream resource.
     *
     * @throws IllegalArgumentException if a template value cannot be sanitized
     */
    @Override
    public Path getStreamStoragePath(Stream stream, Map<String, String> templatingVars) {
        // Copied, not mutated in place: Go mutated the caller's map to inject the default quality.
        var vars = new LinkedHashMap<>(templatingVars);
        vars.putIfAbsent(TemplateConstants.VAR_NAME_QUALITY, VideoConstants.DEFAULT_QUALITY);

        var template = stream.path();

        // Only HLS-only mode uses quality subdirs.
        // DASH-enabled modes (DASH-only or dual) use a flat layout - no quality subdir appended.
        var hlsOnly = stream.distribution().hlsEnabled() && !stream.distribution().dashEnabled();

        var isRootResource = ROOT_RESOURCES.contains(vars.getOrDefault(TemplateConstants.VAR_NAME_RESOURCE, ""));

        if (hlsOnly && !isRootResource && !template.contains(TemplateConstants.QUALITY_VAR)) {
            template += "/" + vars.get(TemplateConstants.VAR_NAME_QUALITY);
        }

        return appPaths.videoDir().resolve(pathTemplateService.replacePlaceholders(template, vars));
    }
}
